### PR TITLE
Hacktoberfest - ParametersDefinition add equals and hashCode

### PR DESCRIPTION
**Pull request by gulyaev13**
Declarative Jenkins Pipelines have to compare ParametersDefinition from job configuration and pipeline for making a decision about updating configuration and save it on disk.

The current implementation compares instances of ParametersDefinition serialized to XMLs because previously `equals()` wasn't implemented in ParametersDefinition implementations. 
https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/pipeline-model-definition-1.7.2/pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy#L738

So this change should improve the performance of starting pipelines.

### Proposed changelog entries

* Entry 1: Add equals and hashCode to ParameterDefinition implementations
* Entry 2: equals working strictly only for the same classes to preventing issues with extended classes

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by daniel-beck** [comment]
Is this filed as a bug? If so, could you provide a reference to it? If not, could you file it?

**Review by MRamonLeon — APPROVED** [review]
It looks to me like an improvement, rather than a bug. I guess we even don't need to change the script referenced because the new `equal` method will do the work and the call to the `isObjectsEqualsXStream` method at https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/pipeline-model-definition-1.7.2/pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy#L742 is not even called.

**Comment by oleg-nenashev** [comment]
Yes, it is not a bug per se. XML comparison is likely to be error-prone, but AFAIK it works.
Regardless of the change type, it is a good performance improvement. Thanks @gulyaev13 !

File: test/src/test/java/hudson/model/ParameterDefinitionTest.java
Comment (jglick) [inline test/src/test/java/hudson/model/ParameterDefinitionTest.java:9]: Since this is a unit test (no `JenkinsRule` dep), it can be moved into the `core` module.

**Comment by gulyaev13** [comment]
@daniel-beck , yes, it isn't a bug, it just a performance improvement.

Originally it was a fix of 2 bugs in pipeline-model-definition:
https://issues.jenkins-ci.org/browse/JENKINS-44809
https://issues.jenkins-ci.org/browse/JENKINS-49500
https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/365

Without equals implementation in ParameterDefenition implementation, I had no way to implement comparison without serialization to XML. But I've added check with equals before XML comparison for better performance ParameterDefenition will implement equals.

File: test/src/test/java/hudson/model/ParameterDefinitionTest.java
Comment (gulyaev13) [reply test/src/test/java/hudson/model/ParameterDefinitionTest.java:9]: @jglick , should I move tests to `core/src/test/java`?

Comment (jglick) [reply test/src/test/java/hudson/model/ParameterDefinitionTest.java:9]: Yes, exactly.

Comment (gulyaev13) [reply test/src/test/java/hudson/model/ParameterDefinitionTest.java]: Done

**Review by res0nance — APPROVED** [review]

**Comment by h1dden-da3m0n** [comment]
Sorry should this be not relevant, but I stumbled over this and added it to my parameter plugin too.
After adding a overwrite for equals and having an all green CI in the PR I had a [SpotBugs](https://ci.jenkins.io/job/Plugins/job/rest-list-parameter-plugin/job/main/39/spotbugs/) error in the main branch CI after the merge.

Thought I should mention this here, but I am unsure if it is even relevant in this context

**Comment by jglick** [comment]
@h1dden-da3m0n

```
io.jenkins.plugins.restlistparam.RestListParameterDefinition.equals(Object) fails for subtypes |  
```

You may want to mark your class `final`.

**Comment by h1dden-da3m0n** [comment]
Hi @jglick .

All I can say is I :man_facepalming: hard at my inability to read apparently (and not noticing that in the first place).

Thanks for the heads up!

**Review by zbynek — CHANGES_REQUESTED** [review]
According to [Object.equals contract](https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#equals(java.lang.Object)) equals should return false when called with `null` argument, but this implementation would throw an exception in such case.

Also the code style is a bit inconsistent here, I'd suggest to always use `if (` rather than `if(`.

**Comment by gulyaev13** [comment]
@zbynek, thanks for your comments!
I've fixed both cases.

File: core/src/main/java/hudson/model/RunParameterDefinition.java
Comment (zbynek) [inline core/src/main/java/hudson/model/RunParameterDefinition.java:221]:
```suggestion
        if (RunParameterDefinition.class != getClass())
```

**Review by zbynek — APPROVED** [review]
One more space, otherwise :+1:

**Comment by res0nance** [comment]
Looks like all conversations are resolved, marking ready for merge and we may merge this in 24h

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Many parameter definition classes are not final. AFAICT, there will be an issue when comparing child classes. Since the logic overrides equals() and hashCode(), child classes will use the same comparison methods unless hashCode/equals are overridden there. If extension classes introduce new fields, it will lead to regressions f the newly introduced fields contain data important for making a decision.

IMHO the logic needs to be updated to delegate to the serialized comparison for if  the methods are called for child classes without overrides

**Comment by gulyaev13** [comment]
@oleg-nenashev, I've handled this case by return false.
Example:
```
if (StringParameterDefinition.class != getClass())
            return false;
```
But I understood that it will break the search in HashMap.

Now I see 3 options:
- Add serialized compassion for `equals()` and `hashCode()` to ParametersDefinition  based on serialized compassion and fallback to that method if called from child classes of implementations without override
- Fallback calls of `equals()` and `hashCode()` to realization in parrent class (Object) in case method called from classes without overrides
- Use serialized compassion in `equals()` only for implementations of ParametersDefinition in case method called from classes without overrides

That you can advise?

**Comment by oleg-nenashev** [comment]
> Add serialized compassion for equals() and hashCode() to ParametersDefinition based on serialized compassion and fallback to that method if called from child classes of implementations without override

Looks like the best way

**Comment by jglick** [comment]
> delegate to the serialized comparison

This sounds like overengineering. If the classes are not `final`, look for _actual_ subclasses out there and fix them to not use subclassing.

**Comment by oleg-nenashev** [comment]
> This sounds like overengineering. If the classes are not final, look for actual subclasses out there and fix them to not use subclassing.

Not the entire Jenkins plugin ecosystem is open source. Many vendors and users create their own components which are not discoverable/fixable by others. Also, even the list on https://www.jenkins.io/doc/developer/extensions/jenkins-core/#parameterdefinition is quite long to recommend that approach

**Comment by gulyaev13** [comment]
@oleg-nenashev , thanks for the comment on the previous case from @jglick. It was quite a blocker for me.
I've made the suggested changes.

**Review by oleg-nenashev — APPROVED** [review]
Should work in all cases though the code became considerably bigger. I think this is the best we are going to get in this logic. Thanks @gulyaev13 for addressing the comments!

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Comment by jglick** [comment]
> Many vendors and users create their own components

My question was about actual known cases of _subclasses_, not just any implementation of `ParameterDefinition`.

**Comment by gulyaev13** [comment]
It would be great to find and add realization of equals() and hashCode() to all known cases of subclasses,  but outside this PR.

**Comment by jglick** [comment]
You missed the point of my comment, but never mind, it can be skipped.

**Comment by oleg-nenashev** [comment]
> My question was about actual known cases of subclasses, not just any implementation of ParameterDefinition.
> You missed the point of my comment, but never mind, it can be skipped.

Yes, known subclasses can be fixed in follow-ups. Sorry for misunderstanding it. This PR should not significantly reduce performance for them right now, since we were serializing XML before for all cases

**Pull request merged**
